### ISSUE TITLE
Cached git sources without specific ref misbehaving on bundler >= 2.5.17

### ISSUE THREAD

**Issue by composerinteralia** [rubygems/rubygems#8046]
### Describe the problem as clearly as you can

On the latest versions of bundler (2.5.17 or higher, specifically after https://github.com/rubygems/rubygems/pull/4469), cached git sources without a specific ref end up misbehaving after a round trip to a remote (e.g. git push, then git pull on a different machine).

This only happens when the `refs` directory of the cached bare git repo doesn't include any files. For example, when first building the cache for my example below it will look something like this (where [d] signifies a directory):

```
vendor/cache/nocturne-7b7b618d6290
├── HEAD
├── config
├── description
├── [d] info
│   └── exclude
├── [d] objects
│   ├── [d] info
│   └── [d] pack
│       ├── pack-d9a59c5435d2b7cc079697c8a7629b538c027499.idx
│       └── pack-d9a59c5435d2b7cc079697c8a7629b538c027499.pack
├── packed-refs
├── [d] refs
│   ├── [d] heads
│   └── [d] tags
└── shallow
```

Note that there are no files inside /refs, so git has nothing to track. After pushing to a remote and pulling down somewhere else, we'll end up with:

```
vendor/cache/nocturne-7b7b618d6290
├── HEAD
├── config
├── description
├── [d] info
│   └── exclude
├── [d] objects
│   └── [d] pack
│       ├── pack-d9a59c5435d2b7cc079697c8a7629b538c027499.idx
│       └── pack-d9a59c5435d2b7cc079697c8a7629b538c027499.pack
├── packed-refs
└── shallow
```

But without a refs directory git doesn't seem to treat this is a git repo. Bundler still operates as if there is an app cache, but the git_proxy starts behaving unexpectedly, running against the outer application git repo instead of the cached one for the gem.

In my example below it ends up hitting https://github.com/rubygems/rubygems/blob/8a079e9061ad4aaf2bc0b9007da8f362b7a2e1f2/bundler/lib/bundler/source/git.rb#L194 and wiping out the cache (the bare check ends up true because it's operating on the outer git repo), then eventually re-fetching to build it again. It loses the `.bundlecache` file along the way.

### Possible solutions

The two solutions I see are

1) Add a `refs/.keep` when building the cache so that directory doesn't go away
2) Repopulate the refs directory when initializing the git proxy

I'm happy to open a PR for either, but wanted to see which solution you prefer (or if you have other ideas). The first one seems slightly easier to me, but I'm not sure if there are any gotchas with adding extra files to a git repo that wouldn't normally be there.

### Did you try upgrading rubygems & bundler?

N/A this is a problem specifically with the new bundler git sources.

### Post steps to reproduce the problem

```
# Init repo
git init
touch .gitignore
echo 'vendor/bundle' >> .gitignore

# Init bundle
bundle init
bundle install
bundle update --bundler
bundle config --local path vendor/bundle
bundle config --local cache_all true

# Cache a gem without a specific ref
bundle add --github composerinteralia/nocturne nocturne --skip-install
bundle cache --no-install

# Simulate pushing to a remote and pulling down on a different machine
git add -A
git clean -dxff vendor

# Fetches again unnecessarily because the cache doesn't look like a git repo
# Also ends up deleting .bundlecache
bundle install --local
```

### Which command did you run?

`bundle install` (with a cached git source gem)

### What were you expecting to happen?

When the bare git repo in the cache includes the refs directory, everything works correctly. `bundle install` works find and doesn't clone or fetch.

### What actually happened?

After a remote round trip, the repo in the cache lost its refs directory and `bundle install` re-fetched everything.

## Environment

```
Bundler       2.5.19
  Platforms   ruby, x86_64-darwin-23
Ruby          3.3.4p94 (2024-07-09 revision be1089c8ec5ba40e09b1553e36b3174bf4014d9d) [x86_64-darwin-23]
  Full Path   /Users/composerinteralia/.asdf/installs/ruby/3.3.4/bin/ruby
  Config Dir  /Users/composerinteralia/.asdf/installs/ruby/3.3.4/etc
RubyGems      3.5.11
  Gem Home    /Users/composerinteralia/.asdf/installs/ruby/3.3.4/lib/ruby/gems/3.3.0
  Gem Path    /Users/composerinteralia/.gem/ruby/3.3.0:/Users/composerinteralia/.asdf/installs/ruby/3.3.4/lib/ruby/gems/3.3.0
  User Home   /Users/composerinteralia
  User Path   /Users/composerinteralia/.gem/ruby/3.3.0
  Bin Dir     /Users/composerinteralia/.asdf/installs/ruby/3.3.4/bin
Tools         
  Git         2.34.1
  RVM         not installed
  rbenv       rbenv 1.2.0
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2024-09-18
Git SHA           d569990361
Released Version  true
```

## Bundler settings

```
cache_all
  Set for your local app (/Users/composerinteralia/code/composerinteralia/deleteme/.bundle/config): true
path
  Set for your local app (/Users/composerinteralia/code/composerinteralia/deleteme/.bundle/config): "vendor/bundle"
```

### Gemfile

```ruby
# frozen_string_literal: true

source "https://rubygems.org"

# gem "rails"

gem "nocturne", "~> 0.0.1", :github => "composerinteralia/nocturne"
```

### Gemfile.lock

```
GIT
  remote: https://github.com/composerinteralia/nocturne.git
  revision: 7b7b618d6290eadf97ff08a6ce4f33c0f2483775
  specs:
    nocturne (0.0.1)
      bigdecimal

GEM
  remote: https://rubygems.org/
  specs:
    bigdecimal (3.1.8)

PLATFORMS
  ruby
  x86_64-darwin-23

DEPENDENCIES
  nocturne (~> 0.0.1)!

BUNDLED WITH
   2.5.19
```

**Comment by deivid-rodriguez** [comment]
Nice catch @composerinteralia!

I agree with everything you said. Option 1 seems easier, but I also don't know of unintended consequences of putting empty `.keep` files in a git bare clone structure.

Alternatively, I wonder if this is just a "problem" for `git rev-parse --is-bare-repository` which is run by that line of code you mentioned. I added that line to properly regenerate caches in the previous format, but we could replace that with something else if everything else works fine, like manually checking for existence of certain files and/or folder.

To sum up, I suggest you try any solution that feels better to you and let's see how it works.

**Comment by composerinteralia** [comment]
> Alternatively, I wonder if this is just a "problem" for git rev-parse --is-bare-repository

I think the problem is that without the refs dir git doesn't see it as a repo at all. E.g. https://github.com/git/git/blob/6531f31ef3bead57a3255fa08efa6e7553c5a9a7/setup.c#L400-L411

**Comment by deivid-rodriguez** [comment]
That makes sense @composerinteralia, it's going to be a problem every time that method is checked, most likely not just when running `--is-bare-repository` 👍.
